### PR TITLE
Allow external triggers to use the cycle point.

### PR DESCRIPTION
**Pull request by trwhitcomb**
While external triggers are handy to avoid polling, the current
specification does not allow any cycle point information to be used
in the external trigger string within a task definition that will be
matched against incoming trigger messages.

This change permits an external trigger to include the current cycle point
(specified as "$CYLC_TASK_CYCLE_POINT") as part of the special tasks
in the suite scheduling definition.  At runtime, the cycle point is used
to perform the match, which means that only a message containing the
correct cycle point will result in the external-trigger dependency being
fulfilled.

### CODE REVIEW

**Comment by hjoliver** [comment]
@trwhitcomb - could you explain the use case for this please?  External triggers are really aimed at integer-cycling suites for processing asynchronous data, where the next cycle point (whatever it is) processes the next dataset.  We don't have cycle points in the trigger strings because the external triggering system should not be expected to know the cycle point of the task that is currently waiting on it.   In a regular date-time cycling suite this feature isn't needed because you can just clock-trigger at the expected data arrival time (and use a few 'retry delays' if the data is a bit late).  But maybe I'm missing something?

**Comment by trwhitcomb** [comment]
@hjoliver absolutely.  It's a bit odd, so I expected some questions!  I'll outline how this came about and hopefully it will answer the use case.

I have two suites running at two difference centers - at center A, I'm running a datetime cycling suite that's packaging a dataset for archival.  At center B,  I'm running an integer cycling suite that downloads  those packages (with a filename determined by the date-time-group, which I use as the external trigger key).  Once that download and archival is complete, I need to signal back to center A that the work is done.  The constraints are that the only connection I have between center A and center B is a message queue and a file transfer connection.

I use two sets of external triggers.  One is the traditional one you mentioned which is running at center B, and looking for new cycle points in the message queue, then kicking off the external trigger with the cycle point as the trigger ID.  The other is used at center A, which is listening to a different message queue, and waits for B to report back that archival is complete so A can clean up the files that were successfully archived.

Unfortunately, the time required for the download (from A to B) as well as the archival to tape (on B) is *very* unpredictable and was going to require a lot of polling.  It seemed more straightforward to be able to use the existing external trigger mechanism.

The difference between this use case (in datetime cycling suites and integer cycling suites) is that I imagine the externally-triggered task is typically the first (or nearly the first) task in a given cycle.  In the datetime suite, I don't need that (because I have better triggers to kick off a cycle!) but I do use it for a later task in the suite that's waiting for an unpredictable-length task to complete.  Using the external triggers allowed me to do this without needing long polling intervals.

Make sense?  Still up to you if you want to merge it, but I found it helpful.

**Comment by hjoliver** [comment]
(@trwhitcomb - I've just returned to NZ from England, will take a look at your explanation as soon as possible...)

**Comment by hjoliver** [comment]
@trwhitcomb - OK I *think* I understand your use case as described.  Basically you have a cycle-point-specific dataset that arrives at a relatively unpredictable wall-clock time, so a clock triggered task that polls (or repeatedly retries) for it isn't ideal.  So you want a task in each cycle point (in the date-time suite) that waits on an external triggering system that does know the specific cycle point of the waiting task.

(My initial thought was just use "cylc trigger" on the waiting task in the right cycle point, but of course we don't have a way of specifying a task that waits indefinitely until someone hits it with "cylc trigger" ... except perhaps a ridiculously large clock-trigger offset, but that seems gratuitous!).

Assuming members of @cylc/core don't see a problem I've missed, I'm happy for it to go in. However, it will need documentation (alongside the existing external trigger documentation, and perhaps along the lines of my explanation above which is shorter and more general than yours?), and an automated test to support it, and a short comment in the source code as well explaining why the cycle point replacement is needed.  (I'll provide some guidance on writing the test shortly...)

**Comment by trwhitcomb** [comment]
@hjoliver You've got it.  I'll need some help with the test writing, though, so will see what you have on that.

**Comment by hjoliver** [comment]
@trwhitcomb  - note I've updated my description (first paragraph of my previous comment) to be more like your use-case - please recheck that I've got it right.

**Comment by hjoliver** [comment]
@trwhitcomb - your code works fine as is, but it would be better to substitute the cycle point once when the task proxy is created, instead of repeatedly on evaluation.  This would also show the concrete cycle point on "view prerequisites", rather than the variable name.   I'll try this, along with a test, on a branch of your branch and post a PR to your branch...

**Comment by hjoliver** [comment]
Tim - see the PR to your branch. Please check my mod works.  Maybe you could take a crack at the documentation?

**Comment by trwhitcomb** [comment]
@hjoliver I really like putting the substitution in the task proxy.  Good catch!  I've merged your PR into mine and confirmed that a) the cycle-point-specific triggering works for my test case and b) everything shows up in the prerequisites view as well.  I've updated the documentation as well - please review.

File: doc/cug.tex [hunk 1]
```diff
@@ -3972,6 +3964,40 @@ \subsubsection{Trigger Types}
 \lstinline=$CYLC_DIR/examples/satellite/ext-triggers/suite.rc= is a working
 example of a simulated satellite processing suite.
 
+External triggers are not normally needed in date-time cycling suites driven
+by real time data that comes in at regular intervals.  In these cases a data
+retrieval task can be clock-triggered (and have appropriate retry intervals
+supplied) to submit at the expected data arrival time, so little time if any
+is wasted in polling.  For tasks that may have a highly variable arrival
```
Comment (hjoliver) [inline doc/cug.tex]: I think "tasks" here is a typo.  Maybe something like this: "However, if the arrival time of the cycle-point-specific data is highly variable, external triggering can be used with the cycle point embedded in the message:"

File: doc/cug.tex [hunk 2]
```diff
@@ -3972,6 +3964,40 @@ \subsubsection{Trigger Types}
...
+\begin{lstlisting}
+# suite "data-proc"
+[scheduling]
+    initial cycle point = 20150125T00
+    final cycle point   = 20150126T00
+    [[special tasks]]
+        external-trigger = variable-wait("data arrived for $CYLC_TASK_CYCLE_POINT")
+    [[dependencies]]
+        [[[T00]]]
+            graph = init-process => variable-wait => post-process
```
Comment (hjoliver) [inline doc/cug.tex]: I'd prefer just `get-data` here, as it is more obviously a "task name" than `variable-wait` (which might be confusing to some users, IMO).

File: doc/cug.tex [hunk 3]
```diff
@@ -3972,6 +3964,40 @@ \subsubsection{Trigger Types}
...
+\end{lstlisting}
+
+Once the variable-length waiting is finished, an external detection system
+should notify the suite like:
```
Comment (hjoliver) [inline doc/cug.tex]: "...like this:"

File: doc/cug.tex [hunk 4]
```diff
@@ -3972,6 +3964,40 @@ \subsubsection{Trigger Types}
...
+\lstset{language=transcript}
+\begin{lstlisting}
+shell$ cylc ext-trigger data-proc "data arrived for 20150126T00" passX12334a
+\end{lstlisting}
+where ``data-proc'' is the suite name, the cycle point has replaced the
+variable in the trigger string, and ``passX12334a'' is the ID string for
+the new event.  The suite passphrase must be installed on triggering account.
```
Comment (hjoliver) [inline doc/cug.tex]: "on *the* triggering account".

**Comment by hjoliver** [comment]
@trwhitcomb - a couple of minor comments on the documentation.  Plus, can you add a note on cycle point substitution in the suite.rc reference for "[scheduling] -> [[special tasks]] -> external-trigger".  Otherwise all good.

@matthewrmshin - can you do or arrange a second review?

**Comment by hjoliver** [comment]
Good, thanks Tim :grin:

**Pull request merged**